patches/minecraft/net/minecraft/world/biome/BiomeForest.java.patch

```diff
@@ -41,18 +33,17 @@
          for (int i = 0; i < 4; ++i)
          {
              for (int j = 0; j < 4; ++j)
-@@ -106,12 +121,12 @@
-                 int l = j * 4 + 1 + 8 + p_185379_2_.nextInt(3);
+@@ -107,11 +120,13 @@
                  BlockPos blockpos = p_185379_1_.func_175645_m(p_185379_3_.func_177982_a(k, 0, l));
  
--                if (p_185379_2_.nextInt(20) == 0)
-+                if (p_185379_2_.nextInt(20) == 0 && net.minecraftforge.event.terraingen.TerrainGen.decorate(p_185379_1_, p_185379_2_, blockpos, net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.BIG_SHROOM))
+                 if (p_185379_2_.nextInt(20) == 0)
++                for (int f_i_o = net.minecraftforge.event.terraingen.TerrainGen.decorate(p_185379_1_, p_185379_2_, p_185379_3_, net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.BIG_SHROOM, 1); f_i_o > 0; f_i_o--)
```

Comment:
The layout makes the logic for this very annyoing to read. Cleanup should be done. It also doesn't actually mimic the real logic. You probably want something like shroom = decorate(shroom, rand(20) == 0 ? 1 : 0; if (shroom > 0) else tree